core/src/class/id.rs

```diff
@@ -1,40 +1,67 @@
+use rquickjs_sys::JSRuntime;
+
 use crate::qjs;
-use std::{cell::Cell, sync::Once};
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        OnceLock, RwLock,
+    },
+};
 
 /// The type of identifier of class
 #[cfg_attr(feature = "doc-cfg", doc(cfg(feature = "classes")))]
 pub struct ClassId {
-    id: Cell<qjs::JSClassID>,
-    once: Once,
+    type_id: AtomicUsize,
 }
 
-unsafe impl Send for ClassId {}
-unsafe impl Sync for ClassId {}
+#[derive(Eq, Hash, PartialEq)]
+struct ClassIdKey(*mut JSRuntime, usize);
+
+unsafe impl Sync for ClassIdKey {}
+unsafe impl Send for ClassIdKey {}
+
+static CLASS_ID_COUNTER: AtomicUsize = AtomicUsize::new(1);
+static CLASS_ID_MAP: OnceLock<RwLock<HashMap<ClassIdKey, qjs::JSClassID>>> = OnceLock::new();
```

Comment:
I rather dislike these globals. If we are going with this approach then I think these should be stored within the runtime opaque so that we don't have to use a RwLock and not have global state.

I also am looking at if it is possible to handle this with a different approach. I find it unfortunate that for every conversion into a rust class we would need to do a HashMap look up for safety. The previous non-ng compatible approach was nice in that it only required a single conditional branch to type check the class. 

We might be able to handle it by having a single QuickJS class for all the rust classes and storing a v-table with a finalizer and tracing method next to the class data. That way checking if something is the right type can be done with checking a single pointer for equality.

Reply:
The CLASS_ID_COUNTER might be fine being a global as I don't really see a way around that in this approach but the map should be stored within the opaque.

Reply:
Found the issue. We're dropping the RT Opaque here: https://github.com/DelSkayn/rquickjs/blob/bb4a97b66cd59c4cbb789b01593f1c195af6abe0/core/src/runtime/raw.rs#L36
We then need the RT Opaque to get the class id in here to run tracer:
https://github.com/DelSkayn/rquickjs/blob/bb4a97b66cd59c4cbb789b01593f1c195af6abe0/core/src/class/ffi.rs#L23
So using Runtime Opaque to store the ClassId wont work. 

> We might be able to handle it by having a single QuickJS class for all the rust classes and storing a v-table with a finalizer and tracing method next to the class data. That way checking if something is the right type can be done with checking a single pointer for equality.

I'm also not a big fan of running a hash table lookup (or worse with locks) everytime we wan't to get class id, but I think using a single QJS class will introduce other restrictions down the line, such as inheritance support etc.

I have to think a bit how this can be done without relying on locking, globals or the runtime Opaque.

Reply:
@saghul it seems like QJS want really designed for this use case where we have the same class being used across multiple runtimes. Whats your take on this?

Reply:
Not sure I got the full context, but I'll try. A given class can be registered with as many runtimes as one wants, but each of them will have a different class ID. If you always register classes in the same order they will coincide, but that's just chance. See: https://github.com/quickjs-ng/quickjs/pull/578

Reply:
Found a simple solution, just drop the Opaque after runtime is dropped.

Reply:
Makes sense! There is now a runtime "atexit" style finalizer, perhaps that can be handy? https://github.com/quickjs-ng/quickjs/blob/d79aa871d1c6c0d3ef315c3e207d519be96e0bf9/quickjs.h#L314